google/cloud/spanner/sql_statement.cc

```diff
@@ -0,0 +1,58 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "google/cloud/spanner/sql_statement.h"
+#include <algorithm>
+
+namespace google {
+namespace cloud {
+namespace spanner {
+inline namespace SPANNER_CLIENT_NS {
+namespace {
+bool CaseInsensitiveStringCompare(std::string const& lhs,
+                                  std::string const& rhs) {
+  if (lhs.size() == rhs.size()) {
+    for (int i = 0; i < static_cast<int>(lhs.size()); ++i) {
+      if (std::tolower(lhs[i]) != std::tolower(rhs[i])) {
```

Comment:
I still believe that case-insensitive is not the right thing, I believe `select * from tbl where foo = 'FOO'` should be different than `select * from tbl where foo = 'foo'`.

Reply:
FYI, here are the [docs](https://cloud.google.com/spanner/docs/lexical#case-sensitivity) on case sensitivity.

Reply:
I have opted to remove equality comparison and make SqlStatement a semi-regular type. In addition to the intricacies of case sensitivity, I find it difficult to wrap my head around whether SqlStatement("select * from foo where colX = 1234) is equal or not equal to SqlStatement("select * from foo where colX = @value", {{"value", Value(1234.0)}}). Until such time as we determine equality is a necessary operator for this type, I would like to leave it out.

Reply:
I can see comparing the *serialized* strings of the statements being a valid use-case (i.e, in unit testing). Comparing the raw strings as-is may not make much sense.

Reply:
It's OK with me to punt on `op==` for now, but if you do, please file a new issue for us to consider adding it later.

Semiregular exists as a concept... because it has to. Semiregular types are still not necessarily something to strive for. It's like a use them when you must, but "regular" is to be preferred. See https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#c11-make-concrete-types-regular